multipass-addressing/src/main/scala/com/precisely/bigdata/sample/spark/MultipassAddressing.scala

```diff
@@ -0,0 +1,198 @@
+
+    // Load the addresses from the csv
+    val addressInputDF = AddressInput.open(session, inputAddressPath)
+
+    val downloadManager = new DownloadManagerBuilder(downloadLocation)
+      .addDownloader(new S3Downloader(session.sparkContext.hadoopConfiguration))
+      .addDownloader(new HDFSDownloader(session.sparkContext.hadoopConfiguration))
+      .addDownloader(new LocalFilePassthroughDownloader())
+      .build()
+
+    // build a AddressingUDF, with custom executor
+    val addressingUdf: UserDefinedFunction = new AddressingBuilder()
+      .withResourcesLocation(resourcesLocation)
+      .withDataLocations(dataLocation)
+      .withDownloadManager(downloadManager)
+      .udfBuilder()
+      .withOutputFields(
+        "location.feature.geometry.coordinates.x as X",
+        "location.feature.geometry.coordinates.y as Y",
+        "address.formattedStreetAddress as formattedStreetAddress",
+        "address.formattedLocationAddress as formattedLocationAddress",
+        "customFields['PRECISION_CODE'] as precisionCode")
+      .withErrorField("error")
+      .forCustomExecutor(new CustomExecutor())
+
+    // call UDF for each row in the address DataFrame, this will result in the dataframe containing a new column for each
+    // of the specified output fields as well as an error column
+    addressInputDF
+      // Adds a new column, represented as a collection comprised of the outputFields and the error field
+      .withColumn("addressing_geocode_result", addressingUdf(map(
+        lit("addressLines[0]"), col("streetName"),
+        lit("city"), col("city"),
+        lit("admin1"), col("state"),
+        lit("postalCode"), col("zip")))
+      )
+      // Persist the addressing result to avoid recalculation when we expand the result
+      .persist()
+      // Expand the result collection such that each output field is a separate column, including the error field.
+      .select("*", "addressing_geocode_result.*").drop("addressing_geocode_result")
+      // Write the dataframe to the specified output folder as csv file
+      .write.mode(SaveMode.Overwrite).option("header", "true").csv(outputDirectory)
+  }
+}
+
+/*
+ * A custom Executor that contains the logic to perform verify and then multipass geocode
+ */
+class CustomExecutor extends AddressingExecutor {
+
+  import com.precisely.addressing.v1.model.Response
+  import com.precisely.addressing.v1.{Addressing, Preferences}
+
+  // set preference to return all fields and Match Mode to Relaxed
+  private val relaxedGeocoderPreferences: Preferences = new PreferencesBuilder()
+    .withReturnAllInfo(true)
+    .withMatchMode("RELAXED")
+    .build()
+
+  override def execute(input: RequestInput, preferences: Option[Preferences], addressing: Addressing): Response = {
+
+    val verifyResponse = addressing.verify(input.requestAddress(), preferences.orNull)
+
+    val inputRequest: RequestAddress = input.requestAddress()
+
+    var multiLineRequest: RequestAddress = inputRequest
```

Comment:
COMMENT: default input for the initial geocode should be the original user specified address to handle the case where verify returns no result